lib/core/src/sync/mod.rs

```diff
@@ -1,49 +1,211 @@
+        let mut client = self.client.lock().await;
+        *client = Some(SyncerClient::connect(self.connect_url.clone()).await?);
+        Ok(())
     }
 
-    async fn apply_changes(&self, changes: &[Record]) -> Result<()> {
-        unimplemented!()
+    async fn listen(self: Arc<Self>) -> Result<()> {
+        let Some(ref mut client) = *self.client.lock().await else {
+            return Err(anyhow!(
+                "Cannot run `get_changes_since`: client not connected"
+            ));
+        };
+
+        let request = ListenChangesRequest::new(utils::now(), self.signer.clone())
+            .map_err(|err| anyhow!("Could not sign ListenChangesRequest: {err:?}"))?;
+        let mut stream = client.listen_changes(request).await?.into_inner();
+
+        let cloned = self.clone();
+        tokio::spawn(async move {
+            while let Some(message) = stream.next().await {
+                match message {
+                    Ok(record) => {
+                        if let Err(err) = cloned.apply_changes(&[record]) {
+                            warn!("Could not apply incoming changes: {err:?}")
+                        };
+                    }
+                    Err(err) => warn!("An error occured while listening for records: {err:?}"),
+                }
+            }
+        });
+
+        Ok(())
     }
 
-    async fn get_changes_since(&self, record_index: u64) -> Result<Vec<Record>> {
-        unimplemented!()
+    fn apply_changes(&self, records: &[Record]) -> Result<()> {
+        let (updatable_records, failed_records) = self.collect_records(records);
+
+        // We persist records which we cannot update (> CURRENT_SCHEMA_VERSION)
+        for record in failed_records {
+            self.persister.insert_record(record)?;
+        }
+
+        // We apply records which we can update (<= CURRENT_SCHEMA_VERSION)
+        for record in updatable_records {
+            if let Err(err) = self.persister.apply_record(record) {
+                warn!("Could not apply record changes: {err:?}");
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn get_changes_since(&self, from_id: i64) -> Result<Vec<Record>> {
+        let Some(ref mut client) = *self.client.lock().await else {
+            return Err(anyhow!(
+                "Cannot run `get_changes_since`: client not connected"
+            ));
+        };
+
+        let request = ListChangesRequest::new(from_id, utils::now(), self.signer.clone())
+            .map_err(|err| anyhow!("Could not sign ListChangesRequest: {err:?}"))?;
+        let records = client.list_changes(request).await?.into_inner().changes;
+
+        Ok(records)
     }
 
     async fn set_record(&self, data: SyncData) -> Result<()> {
-        unimplemented!()
+        let Some(ref mut client) = *self.client.lock().await else {
+            return Err(anyhow!("Cannot run `set_record`: client not connected"));
+        };
+
+        let id = self.persister.get_latest_record_id()? + 1;
```

Comment:
We should have something unique here across devices as the id propagated to the server.

Reply:
If the client is synced, we are sure that the new tip is always unique. There is an issue with concurrency though when, for example, two clients try to send a (different) payment and each calls `set_record` before the change gets caught by the listening stream. I think a trivial fix to this is to simply retry X number of times until it succeeds (perhaps doing this in a separate thread so that it doesn't block the payment response).

Reply:
There are two identifiers that have different purpose:
1. record_id - which is created and set by the client, it shouldn't be synced, it only should be unique and we can use GUID for example. Is is used to identify the record to update/sync.
2. version - which is set by the server and is used to detect changes.

Do you think we can use GUID and not sync this id?

Reply:
> version - which is set by the server and is used to detect changes.

This is set by the client, to show which schema version it is currently using. That was the change we settled on, correct?

> which is created and set by the client, it shouldn't be synced, it only should be unique and we can use GUID for example. Is is used to identify the record to update/sync.

Why do we need a separate id/GUID even? The implementation is currently pretty simple:
- User sends record with an id (local_tip + 1)
- Server checks if given id can be added to the tip (local_tip + 1 == current_tip + 1). If not, return error (client then has to re-sync and try again)
- If it is added, return a response with the new_id (= the same as the one in the request payload, so we update local_tip to local_tip + 1)

This ensures that the client _has to_ be synced, and also that records cannot be added at random.

Reply:
But we don't need to ensure the ientbis synced in order to add a record. Also what do you mean add records at random? Is there any requirement here? 
I am seeing a mechanism of retry on the client that can be avoided by using client generated IDs. Do you see any disadvantage in the suggested approach or any advantage in the current approach?

Reply:
The whole reason behind numbered IDs is that we can fetch from a given height. If we don't use a numbered ID, then we have to create a record timestamp on the server's db and then use that to index.

I agree though that the id in the set record payload is pretty useless and can be avoided. I've re-thought about the use-cases and there seems to be no issue if a client sets a record even when they're behind. 

So I'll change `SetRecordRequest` so it only passes data and version, and keep the `new_id` in `SetRecordResponse` which can then be used by the client to update the local tip.